usecases/Concrete/knowledgeGraph/emodul/emodul_validation.py

```diff
@@ -0,0 +1,57 @@
+import subprocess
+from time import sleep
+from turtle import pos
+import requests
+import os
+import sys
+from pathlib import Path
+from git import Repo
+
+baseDir0 = Path(__file__).resolve().parents[0]
+baseDir1 = Path(__file__).resolve().parents[1]
+baseDir2 = Path(__file__).resolve().parents[2]
+ontologyPath = os.path.join(baseDir2,'ConcreteOntology')
+metadataPath = os.path.join(baseDir0,'E-modul-processed-data/emodul_metadata.csv')
+graphPath = os.path.join(baseDir0,'E-modul-processed-data/EM_Graph.ttl')
+#importedOntologiesPath = os.path.join(baseDir2,'GraphCreation/Ontologies')
+processedDataPath = os.path.join(baseDir0,'E-modul-processed-data')
+
+# clone and run rdfconverter docker
+Repo.clone_from('https://github.com/Mat-O-Lab/RDFConverter.git', os.path.join(baseDir1, 'RDFConverter'))
+subprocess.run(['docker-compose', '-f', os.path.join(baseDir1, 'RDFConverter/docker-compose.yml'), 'up', '-d', 'app'], stderr=subprocess.DEVNULL, stdout=subprocess.DEVNULL)
+
+passed = [] #list of filenames of files that passed validation
+failed = [] #list of tuples (filenames, reports) of files that failed validaton
+
+# read shacl shape
+with open(os.path.join(baseDir1, 'shape.ttl'), 'r') as f:
+    shacl_shape = f.read()
+
+sleep(3)
+s = requests.session()
+
+# iterate over all files ending in .ttl
+for rdf_file in (os.path.join(processedDataPath, f) for f in os.listdir(processedDataPath) if os.fsdecode(f).endswith('.ttl')):
+    with open(rdf_file, 'r') as f:
+        post_data = {'rdf': f.read(), 'shaclShape': shacl_shape}
+    # post files and obtain result of validation
+    r = s.post('http://localhost:8080/rdfconv/validate', files=post_data)
+    if r.text.startswith('VALID'):
+        passed.append(rdf_file)
+    else:
+        failed.append((rdf_file, r.text))
+
+# close session object and shutdown docker container
+s.close()
+subprocess.run(['docker-compose', '-f', os.path.join(baseDir1, 'RDFConverter/docker-compose.yml'), 'down'], stderr=subprocess.DEVNULL, stdout=subprocess.DEVNULL)
+
+# print passed files
+print('files that passed shacl validation:')
```

Comment:
There should not be a print, but a check for files that have successfully passed the pytest.

Reply:
Hi @joergfunger, about "check for files that have successfully passed the pytest"
Could you please give an example of the output or something else that satisfies you?

Reply:
So of I write a standard test that returns a result (e.g. that there is not violation of the constraints), then we should check for this in the test (so not just printing the result, but adding assertions https://docs.pytest.org/en/6.2.x/example/simple.html). Same when something might go wrong, so the constraints are not fulfilled. The function then returns the failed constraints, and we can add assertions such that the test checks that exactly these failed constraints (for a given data set that does not fulfill the constraints) are returned.

Reply:
Since this is a test environment, the dataset does not change. This is not supposed to be used dynamically for any data set, but this workflow should rather be a workflow test that demonstrates the features.

Reply:
https://github.com/BAMresearch/ModelCalibration/blob/shacl-validation-action/usecases/Concrete/knowledgeGraph/emodul/emodul_validation.py#L101-L107

Thank you, this explanation helped. So I've changed a lot of the validation code now and we can now test that a specific shape of the shapes graph has failed or not. As shown in the lines of code linked above. We can then define shapes in the shapes graph that should fail or that should not fail and check that they do (or don't) via pytest.